File: Data Structures & Algorithms/top-k-elements-in-list/submission-4.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        List<Integer>[] bucket = new List[nums.length+1];
        for(int i : nums){
            map.put(i, map.getOrDefault(i,0)+1);
        }
        for(int i : map.keySet()){
            int freq = map.get(i);
            if(bucket[freq]== null){
                bucket[freq] = new ArrayList<>();
            }
            bucket[freq].add(i);        
            }
            int [] res = new int[k];
            int index =0;
            for(int i = bucket.length-1; i >0 && index<k ; i--){
                if(bucket[i]!= null){ 
                for(int n : bucket[i]){
                    res[index++]=n;
                    if(index==k){
                        return res;
                    }
                }
            }
            }
            return res;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prod,zerocnt = 1, 0
        for n in nums: 
            if n:
                prod*=n
            else:
                zerocnt+=1

        if zerocnt>1: return [0]*len(nums)
        res=[1]*len(nums)
        for i , n in enumerate(nums): 
            if zerocnt: res[i]=0 if n else prod
            else:
                res[i]=prod//n

        return res